src/python/pants/source/payload_fields.py

```diff
@@ -54,12 +61,14 @@ def source_paths(self):
 
   @property
   def address(self):
-    """Returns the address this sources field refers to (used by some derived classses)"""
+    """Returns the address this sources field refers to (used by some derived classes)"""
     return self._ref_address
 
   def has_sources(self, extension=None):
     if not self.source_paths:
       return False
+    if not extension:
```

Comment:
The logic is actually `if source_paths and not extension`; ie at least 1 source and extension doesn't matter - so not so odd? Agreed its hard to read, but it is now tested at least.